File: 2444-longest-ideal-subsequence/2444-longest-ideal-subsequence.java
class Solution {
    public int longestIdealString(String s, int k) {
        int[] dp = new int[26];
        int res = 1;

        for(char c: s.toCharArray()) {
            int curr = c - 'a';
            for(int i = Math.max(0, curr - k); i <= Math.min(25,curr+k); i++) dp[curr] = Math.max(dp[curr], dp[i]);
            res = Math.max(++dp[curr], res);
        }

        return res;
    }
}

File: 2444-longest-ideal-subsequence/2444-longest-ideal-subsequence.cpp
class Solution {
public:
    int longestIdealString(string s, int k) {
        int dp[26] = {0}, res = 1;
        
        for(char &c: s) {
            int curr = c - 'a';
            for(int i = max(0, curr - k); i <= min(25,curr + k); i++) dp[curr] = max(dp[curr], dp[i]);
            res = max(res, ++dp[curr]);
        }

        return res;
    }
};